op-chain-ops/interopgen/deploy.go

```diff
@@ -0,0 +1,322 @@
+package interopgen
+
+import (
+	"errors"
+	"fmt"
+	"math/big"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/ethereum/go-ethereum/log"
+
+	"github.com/ethereum-optimism/optimism/op-chain-ops/foundry"
+	"github.com/ethereum-optimism/optimism/op-chain-ops/genesis"
+	"github.com/ethereum-optimism/optimism/op-chain-ops/genesis/beacondeposit"
+	"github.com/ethereum-optimism/optimism/op-chain-ops/interopgen/deployers"
+	"github.com/ethereum-optimism/optimism/op-chain-ops/script"
+)
+
+var (
+	// sysGenesisDeployer is used as tx.origin/msg.sender on system genesis script calls.
+	// At the end we verify none of the deployed contracts persist (there may be temporary ones, to insert bytecode).
+	sysGenesisDeployer = common.Address(crypto.Keccak256([]byte("System genesis deployer"))[12:])
+)
+
+func Deploy(logger log.Logger, fa *foundry.ArtifactsFS, srcFS *foundry.SourceMapFS, cfg *WorldConfig) (*WorldDeployment, *WorldOutput, error) {
+	// Sanity check all L2s have consistent chain ID and attach to the same L1
+	for id, l2Cfg := range cfg.L2s {
+		if fmt.Sprintf("%d", l2Cfg.L2ChainID) != id {
+			return nil, nil, fmt.Errorf("chain L2 %s declared different L2 chain ID %d in config", id, l2Cfg.L2ChainID)
+		}
+		if !cfg.L1.ChainID.IsUint64() || cfg.L1.ChainID.Uint64() != l2Cfg.L1ChainID {
+			return nil, nil, fmt.Errorf("chain L2 %s declared different L1 chain ID %d in config than global %d", id, l2Cfg.L1ChainID, cfg.L1.ChainID)
+		}
+	}
+
+	deployments := &WorldDeployment{
+		L2s: make(map[string]*L2Deployment),
+	}
+
+	l1Host := createL1(logger, fa, srcFS, cfg.L1)
+	if err := l1Host.EnableCheats(); err != nil {
+		return nil, nil, fmt.Errorf("failed to enable cheats in L1 state: %w", err)
+	}
+
+	l1Deployment, err := initialL1(l1Host, cfg.L1)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to deploy initial L1 content: %w", err)
+	}
+	deployments.L1 = l1Deployment
+
+	superDeployment, err := deploySuperchainToL1(l1Host, cfg.Superchain)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to deploy superchain to L1: %w", err)
+	}
+	deployments.Superchain = superDeployment
+
+	for l2ChainID, l2Cfg := range cfg.L2s {
+		l2Deployment, err := deployL2ToL1(l1Host, cfg.Superchain, superDeployment, l2Cfg)
+		if err != nil {
+			return nil, nil, fmt.Errorf("failed to deploy L2 %d to L1: %w", &l2ChainID, err)
+		}
+		deployments.L2s[l2ChainID] = l2Deployment
+	}
+
+	out := &WorldOutput{
+		L2s: make(map[string]*L2Output),
+	}
+	l1Out, err := completeL1(l1Host, cfg.L1)
```

Comment:
We deploy all contracts to the L1 because we need to compute the genesis block hash to put into the L2 genesis configs, and can thus not mutate the L1 state after creating the final config for any particular L2. Will add comments.